### DISCUSSION TITLE
Using useAsyncData in composables

### DISCUSSION THREAD

**Discussion by itzaks** [nuxt/nuxt#16053] [Questions] [answered]
I've tried using useAsyncData inside composables, judging from a previous discussion I posted: https://github.com/nuxt/framework/discussions/3579. But I recently ran into a lot of issues when nesting composables, and awaiting async composables. I've also understood that this isn't really recommended.

Here's some example code to visualize my structure.

````ts
const useUser = async () => {
  const { data: user } = await useAsyncData('user', () => gqlUser())
  return { user }
}

const usePosts = async () => {
  const { data: posts } = await useAsyncData('posts', () => gqlPosts())
  return { posts }
}

const useBlog = async () => {
  const [{user}, {posts}] = await Promise.all([useUser(), usePosts()])
  const userAndPost = computed(() => user.value + posts.value)
}

const useSomethingElse = async () => {
  const { user } = await useUser()
}
````

However, this has led up to a lot of different issues, both hydration mismatches and nuxt instance unavailable, because of this, I guess – https://github.com/nuxt/framework/issues/5740#issuecomment-1229197529.

Also, as in my example, if one composable with useAsyncData is invoked from more than one place, and the asyncdata promise hasn't yet resolved, it will trigger multiple fetches. I tried putting them into `createGlobalState` from `vueUse` but it just created more issues.

It just seems this is not the way to do it, however the ssr fetch, caching and the refresh mechanics is just really nice to have. So to me it is a bummer not to be able to use it in composables. Has anyone else found a way to actually use useAsyncData together with composables in this way?

**Comment by Gringo675** [comment]
Yeah, have the same feeling. It’s pain to work with useAsyncData in composables. Should use callWithNuxt helper, unwatch watchers in onUnmounted hook, keep in mind that if it is lazy, it resolves only in onBeforeMounted hook and so on. It works, but looks rather messy. So in the end I decided to use component instead of composable. It is not so flexible, but covered my needs.

**Reply by itzaks** [reply]
Right.. that is one way to do it, but in my current project, we rely a lot on a bunch of computed stuff in composables, based on the async data. Moving the fetch to the component and passing it into the composables just feels a bit cumbersome.

**Comment by IsraelOrtuno** [comment]
I am trying to debug something and may be related to this. Do you find this issue even if you do not nest composables? In your case using `useUser` or `usePosts` directly from a page/component.

**Reply by itzaks** [reply]
Not really, seems like using async composables is gist of the issue, and also trying to "help" with createGlobalState which just complicates things with regards to nuxt context and SSR-rendering. If I call it and await only from the component, it seems to work fine.

**Comment by itzaks** [comment]
I'm moving away from that solution now, and currently I'm looking at something with useAsyncData and `immediate: false`, and manually fetching from the composables like this.


````ts
const useUser = () => {
  const { data: user, execute: fetchUser} = useAsyncData('user', () => 
     gqlUser(), 
  { immediate: false })

  return { user, fetchUser }
}
....

const useBlog = async () => {
  const {user} = useUser()
  const {posts} = usePosts()

  const userAndPost = computed(() => user.value + posts.value)
}
````
And then in the component:

````ts
<template lang="ts" setup>
const {fetchUser} = useUser()
const {fetchPosts} = usePosts()
await Promise.all([fetchUser(), fetchPosts()])
</template>
````

However the downside here, is that if I might need to fetch the data elsewhere, in another component, I lose the caching mechanism of await-ing useAsyncData straight away. Manually calling means it always fetches the data every time...

**Reply by IsraelOrtuno** [reply]
How do you get that caching mechanism? I was able to do it in previous RC releases but since v3 stable release `initialCache` was removed and data is fetched every time.

Also, do you think this may be related of how composables work? I am experiencing some issues when deploying in Vercel just like "Nuxt Instance unavailable" and some others which just do not provide enough feedback to start debugging but can't really reproduce them, kind of random. I of do the same as you using the composables for data fetching (I don't do any useAsyncData there), they basically fetch the data and map the results to the frontend models.

**Reply by itzaks** [reply]
Yeah, the above solution didn't really work out in the end, and I ended up using this: https://github.com/nuxt/framework/discussions/9432#discussioncomment-4315113

**Comment by itzaks** [comment, marked as answer]
I ended up creating my own composable for this, that is a lightweight version inspired by useAsyncData, which sets data on useState and exposes two methods, execute and refresh. `execute` will work as an initial fetch, supposed to run in initial setups, and then `refresh` is supposed to be run when you want to fetch new data.

You can check it out here if anyone else is interested of this solution:
https://gist.github.com/itzaks/3ef8a8003abde0c9d72a19cccfd7d454
